Queries (#3)
temporalio/edu-interacting-with-workflows-typescript: querying-your-workflows/sending-queries.md

## Sending a Query with the SDK

> After defining and handling Queries in your Workflow, the next step is to send Queries, which are sent from a Temporal Client. The SDK provides the `WorkflowHandle.query` method to query a running or completed Workflow.

Author: fixed!

```typescript
import { Client } from "@temporalio/client";
import { getValueQuery } from "./workflows";

async function run(): Promise<void> {
  const client = new Client();
  const handle = client.workflow.getHandle("state-id-0");
  const meaning = await handle.query(getValueQuery, "meaning-of-life");
  console.log({ meaning });
}
```

In the example above, we use the `getHandle` method to obtain a handle to the Workflow instance of `state-id-0`. The `query` method is then called with this handle to send the `getValueQuery` with an argument `meaning-of-life`.

## Sending a Query with the CLI

> You can also use the `temporal` CLI to send a Query to a Workflow Execution. To send a Query with the Temporal CLI, you use the `temporal workflow query` or command.

Author: Fixed

## Sending a Query with the SDK

> After defining and handling Queries in your Workflow, the next step is to send a Query, which is sent from a Temporal Client. The SDK provides the `WorkflowHandle.query` method to query a running or completed Workflow.

Reviewer: This sounds weird:

```
After defining and handling Queries in your Workflow
               ^^^^^^^^^^^^
```

```typescript
import { Client } from "@temporalio/client";
import { getValueQuery } from "./workflows";

async function run(): Promise<void> {
  const client = new Client();
  const handle = client.workflow.getHandle("state-id-0");
  const meaning = await handle.query(getValueQuery, "meaning-of-life");

>   console.log({ meaning });

Reviewer: ```suggestion
  console.log(meaning);
```

}
```

> In the example above, we use the `getHandle` method to obtain a handle to the Workflow instance of `state-id-0`. The `query` method is then called with this handle to send the `getValueQuery` with an argument `meaning-of-life`.

Reviewer: ```suggestion
In the example above, we use the `getHandle` method to obtain a handle to the Workflow Execution with ID `state-id-0`. The `query` method is then called on this handle to send the `getValueQuery` to that Workflow, with an argument `meaning-of-life`.
```

> You'll notice that a Query does not add any Events into your Event History, and the output will be just logged in your CLI.

Reviewer: ```suggestion
You'll notice that a Query does not add any Events into your Event History. The return value of the Query will simply be logged to the console.
```

## Sending a Query with the CLI

> You can also use the Temporal CLI to send a Query to a Workflow Execution. To send a Query with the Temporal CLI, you use the `temporal workflow query` or command.

Reviewer: ```suggestion
You can also use the Temporal CLI to send a Query to a Workflow Execution. To send a Query with the Temporal CLI, you use the `temporal workflow query`command.
```

> Let's say that we have a Workflow that adds members into a group - with a `addMembersIntoGroups` Workflow ID - and we want to query for a member's information using the Query Type `getMemberInfo`. The member that we want to query for is "angela". Our command will look like this:

Reviewer: > we have a Workflow that adds members into a group - with a `addMembersIntoGroups` Workflow ID

That's a surprising example. I honestly don't understand the use case that would justify what is described here, and that makes it difficult to make sense of the snippet.

May I suggest instead to reuse the same workflow/query/args/return as in the SDK example above?

```bash

> $ temporal workflow query \

Reviewer 2: Please indent the continued lines in this command by four spaces.

Author: added

  --workflow-id="addMembersIntoGroups" \
  --type="getMemberInfo" \
  --input=\"angela\"
```

> Upon successful execution, the CLI confirms with `Query result as JSON: 123`.

Reviewer 2: I don't have the code to try this myself, but you should doublecheck whether the output you mention here is accurate. The output I saw during the previous exercise was formatted differently:

```
Query result:
["\"Z1238\""]
```

Author: fixed
